File: nodejs/app/stores/WikipediaStore.js
var WikipediaCollections 	= require('../collections/WikipediaCollections.js');

var collections = undefined;


var createUpdatedObject = function(category, title, day, time, year) {
	var updatedObject = {
		category: category,
		title: title,
		day: day,
		updated: time
	};
	if(year) {
		updatedObject['year'] = year;
	}
	return updatedObject;
}

// create query object for find
var createQueryObject = function(day, year, keyword) {
	var queryObj = {};
	if (day) {
		queryObj['day'] = day;
	}
	if (year) {
		queryObj['year'] = year;
	}
	if (keyword) {
		queryObj['title'] = {"$regex": keyword , "$options": "i" };
	}
	return queryObj;
};


// Wikipedia store, the first layer over database
var wikipediaStore = {
	// insert database
	insertInCategory: function(category, title, day, time, year) {
		var updatedObject = createUpdatedObject(category, title, day, time, year);

		// if year exists then it's not a holiday
		if (year) {
			return collections[category].update({
				title: title,
				category: category
			},
			updatedObject,
			{
				upsert: true
			}
		);
		} else {
			// it's an event, death or birth
			return collections[category].update({
				title: title,
				category: category
			},
			updatedObject,
			{
				upsert:true
			}
		);
		}
	},

	// find for a specific category
	findInCategory: function(category, day, year, keyword) {
		var queryObj = createQueryObject(day, year, keyword);
		console.log(category);
		if (category === 'holidaysandobservances') {
			delete queryObj.year;
		}
		return collections[category].find(queryObj).toArray();
	},

	// find in all categories
	findAll: function(day, year, keyword) {
		var queryObj = createQueryObject(day, year, keyword);
		var results = [];
		var promises = [];
		return new Promise(function(resolve, reject) {
			for (var key in collections) {
				if (key !== 'holidaysandobservances') {
					promises.push(collections[key].find(queryObj).toArray());
				}
				else {
					// remove property of year, if exists
					if (queryObj.year) {
						delete queryObj.year;
					}
					promises.push(collections[key].find(queryObj).toArray());
				}
			}
			Promise.all(promises)
				.then(function(data) {
					for (var index in data) {
						results = results.concat(data[index]);
					}
					resolve(results);
				})
				.catch(function(err) {
					reject(err);
				})
		});
	},
	cleanUpCategory: function (category, oldTime) {
		return new Promise(function(resolve, reject) {
			collections[category.toLowerCase()].deleteMany(
				{ updated: { "$lte" : oldTime} }
			)
				.then(function(results) {
					resolve(results.deletedCount);
				})
				.catch(function(err){
					reject(err);
				});
		});
	}
};


module.exports = function(dbInstance) {
	if (dbInstance && !collections) {
		collections = WikipediaCollections(dbInstance);
	}

	return wikipediaStore;
};

File: python/app/stores/wikipediaStore.py
from app.databases.database import Database

"""
Create query object used by find function
@:param query parameters (day, year, keyword)
@:return dictionary of query params
"""
def createQueryObject(day, year, keyword):
    queryObject = {}
    if day is not None:
        queryObject['day'] = day

    if year is not None:
        queryObject['year'] = year


    if keyword is not None:
        queryObject['title'] = { '$regex' : keyword }

    return queryObject


"""
Convert _id from object to string
@:param dictionary from DB
@:return dictionary with _id changed
"""
def convertId(document):
    document['_id'] = str(document['_id'])
    return document


"""
Database layer interface. (database store)
Expose functions for services.
saveEntry - add a new entry to a category
findInCategory - find documents that match parameters in categories
removeOldData - remove old data from database
"""
class WikipediaStore:
    def __init__(self):
        self.collections = Database(None, None).wikiCollections

    def __str__(self):
        print(self.collections)

    """
    @:param query params from request
    @:return True if files were updated
            String (id) if entry is added
            False otherwise
    """
    def saveEntry(self, title, day, category, time, year):
        if not title or not day or not category:
            return None
        result = None
        if year is not None:
            result = self.collections[category].update_one(
                {
                    'title': title,
                    'category': category
                },
                {
                    '$set' : {
                        'title': title,
                        'day': day,
                        'year': year,
                        'category': category,
                        'updated': time
                    }
                },
                True  #upsert set to true
            )
        else:
            result = self.collections['holidaysandobservances'].update_one(
                {
                    'title': title,
                    'category': category
                },
                {
                    '$set': {
                        'title': title,
                        'day': day,
                        'year': year,
                        'category': category,
                        'updated': time
                    }
                },
                True  # upsert set to true
            )

        if result.acknowledged == True:
            if result.matched_count > 0 and result.modified_count > 0:
                return True
            elif result.matched_count == 0 and result.upserted_id:
                return result.upserted_id
            else:
                return False


    """
    Find in a specific category
    @:param query params for find function
    @:return a list of documents from database
    that matches query params
    """
    def findInCategory(self, category, day, year, keyword):
        results = []
        queryObj = createQueryObject(day, year, keyword)
        if category == 'holidaysandobservances':
            del queryObj['year']
        return list(map(convertId, [doc for doc in self.collections[category].find(queryObj)]))

    """
    Find in all categories
    @:param query params for find function
    @:return a list of documents from database
    that matches query params
    """

    def findAll(self, day, year, keyword):
        result = []
        queryObj = createQueryObject(day, year, keyword)
        for category in self.collections:
            if category != 'holidaysandobservances':
                result += list(map(convertId, [doc for doc in self.collections[category].find(queryObj)]))
            else:
                queryObj2 = queryObj.copy()
                del queryObj2['year']
                result += list(map(convertId, [doc for doc in self.collections[category].find(queryObj)]))
        return result


    """
    @param: last time when databse was fetched
    """
    def removeOldData(self, oldTime):
        pass